src/util.rs

```diff
@@ -91,6 +91,68 @@ impl Drop for StructWriter {
     }
 }
 
+pub(crate) struct StructReader {
+    start: *mut u8,
+    offset: usize,
+    layout: Layout,
+}
+
+impl StructReader {
+    /// Creates a struct reader that has the given initial capacity `capacity`,
+    /// and whose allocation is aligned to `align`
+    pub fn new(capacity: usize, align: usize) -> Self {
+        let layout = Layout::from_size_align(capacity, align).unwrap();
+        let start = unsafe { std::alloc::alloc(layout) };
+        Self {
+            start,
+            offset: 0,
+            layout,
+        }
+    }
+
+    /// Reads a given type from the internal buffer.
+    /// This advances the internal pointer by the size of the read type, such that a given instance of
+    /// the given type can only be read once.
+    ///
+    /// # Safety: TODO
```

Comment:
We can re-use the safety and panic docs for StructWriter:
    /// # Safety:
    /// The caller must ensure the internal pointer is aligned suitably for reading a T.
    /// In most C APIs (like Wireguard NT) the structs are setup in such a way that calling read
    /// repeatedly to read packed data always yields a struct that is aligned because the
    /// previous struct was aligned.
    ///
    /// # Panics
    /// 1. If reading a struct of size T would overflow the buffer.
    /// 2. If the internal pointer does not meet the alignment requirements of T.

The safety section is a little chunky so feel free to rewrite if you can find a better way to express it.